File: backend/app/schemas/document.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# Status values must match DocumentStatus enum exactly
DocumentStatusLiteral = Literal["uploaded", "processing", "indexed", "failed", "inactive"]


class DocumentResponse(BaseModel):
    """Full document metadata returned by all admin document endpoints."""

    model_config = {"from_attributes": True}

    id: int
    filename: str
    original_filename: str
    content_type: str
    file_size: int
    page_count: int | None
    document_id: str
    status: DocumentStatusLiteral
    error_message: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime
    ingested_at: datetime | None


class DocumentListResponse(BaseModel):
    documents: list[DocumentResponse]
    total: int


class DocumentStatusUpdateRequest(BaseModel):
    """Request body for PATCH /admin/documents/{id}/status."""

    is_active: bool = Field(
        description="Set to false to deactivate, true to reactivate the document."
    )


class DocumentChunkStats(BaseModel):
    """Qdrant vector counts for a single document."""

    document_id: str
    chunk_count: int
    collection_total: int


class KnowledgeBaseStats(BaseModel):
    """Aggregate knowledge-base health stats for the admin dashboard."""

    total_documents: int
    by_status: dict[str, int]
    total_indexed_chunks: int
    qdrant_reachable: bool

File: frontend/src/types/document.ts
export type DocumentStatus = "uploaded" | "processing" | "indexed" | "failed" | "inactive";

export interface GovernmentDocument {
  id: number;
  filename: string;
  original_filename: string;
  content_type: string;
  file_size: number;
  page_count: number | null;
  document_id: string;
  status: DocumentStatus;
  error_message: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  ingested_at: string | null;
}

export interface DocumentListResponse {
  documents: GovernmentDocument[];
  total: number;
}

export interface DocumentStatusUpdateRequest {
  is_active: boolean;
}

export interface DocumentChunkStats {
  document_id: string;
  chunk_count: number;
  collection_total: number;
}

export interface KnowledgeBaseStats {
  total_documents: number;
  by_status: Record<DocumentStatus, number>;
  total_indexed_chunks: number;
  qdrant_reachable: boolean;
}
